src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -1146,15 +1147,88 @@ MsalIdTokenCacheItem ITokenCacheInternal.GetIdTokenCacheItem(MsalAccessTokenCach
                 tenantProfiles?.Values);
         }
 
+        async Task ITokenCacheInternal.StopLongRunningOboProcessAsync(string longRunningOboCacheKey, AuthenticationRequestParameters requestParameters)
+        {
+            requestParameters.RequestContext.Logger.Verbose(() => $"[StopLongRunningOboProcessAsync] Entering token cache semaphore. Count {_semaphoreSlim.GetCurrentCountLogMessage()}");
+            await _semaphoreSlim.WaitAsync(requestParameters.RequestContext.UserCancellationToken).ConfigureAwait(false);
+            requestParameters.RequestContext.Logger.Verbose(() => "[StopLongRunningOboProcessAsync] Entered token cache semaphore");
+
+            try
+            {
+                requestParameters.RequestContext.Logger.Info("[StopLongRunningOboProcessAsync] Stopping long running OBO process by removing tokens from cache.");
+
+                ITokenCacheInternal tokenCacheInternal = this;
+
+                try
+                {
+                    if (tokenCacheInternal.IsAppSubscribedToSerializationEvents())
+                    {
+                        var args = new TokenCacheNotificationArgs(
+                            tokenCache: this,
+                            clientId: ClientId,
+                            account: null,
+                            hasStateChanged: true,
+                            tokenCacheInternal.IsApplicationCache,
+                            suggestedCacheKey: longRunningOboCacheKey,
+                            hasTokens: tokenCacheInternal.HasTokensNoLocks(),
+                            suggestedCacheExpiry: null,
+                            cancellationToken: requestParameters.RequestContext.UserCancellationToken,
+                            correlationId: requestParameters.RequestContext.CorrelationId,
+                            requestScopes: requestParameters.Scope,
+                            requestTenantId: requestParameters.AuthorityManager.OriginalAuthority.TenantId,
+                            identityLogger: requestParameters.RequestContext.Logger.IdentityLogger,
+                            piiLoggingEnabled: requestParameters.RequestContext.Logger.PiiLoggingEnabled);
+
+                        await tokenCacheInternal.OnBeforeAccessAsync(args).ConfigureAwait(false);
+                        await tokenCacheInternal.OnBeforeWriteAsync(args).ConfigureAwait(false);
+                    }
+
+                    RemoveOboTokensInternal(longRunningOboCacheKey, requestParameters.RequestContext);
+                }
+                finally
+                {
+                    if (tokenCacheInternal.IsAppSubscribedToSerializationEvents())
+                    {
+                        var args = new TokenCacheNotificationArgs(
+                           tokenCache: this,
+                           clientId: ClientId,
+                           account: null,
+                           hasStateChanged: true,
```

Comment:
How do you know the state changed?

Reply:
This same value is present in the remove account method below and it looks like it assumes there will be a change in state. not sure if entirely correct actually, this notification args are being senf in the befrore access methods so there should be a state change yet, but I was confused as to why the remove account logic is doing this. Is this a bug in the remove account method?